operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+}
+
+// WatchResources configures resource watching for the given cluster, including StatefulSets and other resources.
+func (r *ResourceClient[T, U]) WatchResources(builder Builder, cluster U) error {
+	// set that this is for the cluster
+	builder.For(cluster)
+
+	// set an Owns on node pool statefulsets
+	builder.Owns(&appsv1.StatefulSet{})
+
+	for _, resourceType := range r.simpleResourceRenderer.WatchedResourceTypes() {
+		mapping, err := getResourceScope(r.mapper, r.scheme, resourceType)
+		if err != nil {
+			return err
+		}
+
+		if mapping.Name() == meta.RESTScopeNamespace.Name() {
+			// we're working with a namespace scoped resource, so we can work with ownership
+			builder.Owns(resourceType)
+			continue
+		}
+
+		// since resources are cluster-scoped we need to call a Watch on them with some
+		// custom mappings
+		builder.Watches(resourceType, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			if owner := r.ownershipResolver.OwnerForObject(o); owner != nil {
+				return []reconcile.Request{{
+					NamespacedName: *owner,
+				}}
+			}
+			return nil
+		}))
+
+	}
+
+	return nil
+}
+
+// DeleteAll deletes all resources owned by the given cluster, including node pools.
+func (r *ResourceClient[T, U]) DeleteAll(ctx context.Context, owner U) (bool, error) {
+	// since this is a widespread deletion, we can delete even stateful sets
+	resources, err := r.listAllOwnedResources(ctx, owner, true)
+	if err != nil {
+		return false, err
+	}
+
+	alive := []client.Object{}
+	for _, o := range resources {
+		if o.GetDeletionTimestamp() == nil {
+			alive = append(alive, o)
+		}
+	}
+
+	// attempt to delete as many resources in one pass as we can
+	errs := []error{}
+	for _, resource := range alive {
+		if err := r.client.Delete(ctx, resource); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return len(alive) > 0, errors.Join(errs...)
+}
+
+// listResources lists resources of a specific type and object, returning them as an array.
+func (r *ResourceClient[T, U]) listResources(ctx context.Context, object client.Object, opts ...client.ListOption) ([]client.Object, error) {
+	kind, err := getGroupVersionKind(r.client.Scheme(), object)
+	if err != nil {
+		return nil, err
+	}
+
+	olist, err := r.client.Scheme().New(schema.GroupVersionKind{
+		Group:   kind.Group,
+		Version: kind.Version,
+		Kind:    kind.Kind + "List",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("initializing list: %w", err)
+	}
+	list, ok := olist.(client.ObjectList)
+	if !ok {
+		return nil, fmt.Errorf("invalid object list type: %T", object)
+	}
+
+	if err := r.client.List(ctx, list, opts...); err != nil {
+		return nil, fmt.Errorf("listing resources: %w", err)
+	}
+
+	converted := []client.Object{}
+	items := reflect.ValueOf(list).Elem().FieldByName("Items")
+	for i := 0; i < items.Len(); i++ {
+		item := items.Index(i).Addr().Interface().(client.Object)
+		item.GetObjectKind().SetGroupVersionKind(*kind)
```

Comment:
Isn't GVK guaranteed to be populated when reading from the k8s API?

Reply:
Surprisingly, not when doing a `List` operation -- at least that's what tests uncovered, but it could also be an `envtest` thing?

Reply:
HUH! TIL. `envtest` (AFAIK) runs the literal kube-api server. There shouldn't be any distinction between the API semantics.

Mind calling that out in a comment so we (I) don't forget it?